### DISCUSSION TITLE
Guidance on NHSUK Frontend library with Java/Thymeleaf

### DISCUSSION THREAD

**Discussion by yantantether** [nhsbsa/nhsbsa-digital-playbook#290] [Development]
Projects using Node.js for frontends have an advantage over Java/Thymeleaf projects, that they can use a direct NPM dependency to the frontend library and include Nunjucks components. This means they can easily get the latest version of HTML, CSS and JS.

### CSS/JS

Projects using Java/Thymeleaf have adopted different approaches to re-use the library's precompiled assets (CSS and JS only):

* Add the frontend library Webjar as a Maven dependency
* Use NPM/Grunt to pull and compile the dependency as a build step
* Commit the pre-compiled assets into the project source

We should agree on, and advise on the best approach.

### HTML

We don't have a Thymeleaf port of the frontend nunjucks library. This means that projects have to manage their own HTML. If the HTML for a nunjucks component changes, projects need to track these changes down into their own codebase. There isn't an easy way to do this, and it is likely that we have missed changes.

Potential options:

* Create a Thymeleaf port of the library
* Have a team responsible for picking out changes in HTML and communicating to project teams

We should agree on, and advise on the best approach.

**Comment by nhsbsa-wicoo** [comment]
Yeah I'd like a Thymeleaf port of the library, but if that's too time consuming, would be happy communicate out the HTML changes

**Comment by nhsbsa-jamcn** [comment]
I started work on a proof-of-concept library recently (available internally on Gitlab; nhsuk-frontend-thymeleaf). So far it can provide the static content (scripts and styles) from the toolkit in addition to a few components in the form of fragments - this all works fine so far. The component fragments were tricky to implement as they needed to be flexible, so currently require a lot of arguments. Providing models in Java to represent these fragments to make it easier for developers to set them up could help with this.

The real challenge is going to be how we define the fragments (and supporting models) and keep them inline with the toolkit. One possibility could be to set up some scripts as a form of pipeline which pulls the nunjucks macro files from the toolkit repository, renders them into HTML, replacing the variables expected with placeholders for ThymeLeaf variables. I can't see this being straightforward to implement, so an alternative may be preferable.

I was looking into whether we could make Spring first render a page with Thymeleaf, then with nunjucks. This would be ideal, as then we could just keep existing ThymeLeaf-based projects the same, but then projects could start gradually substituting HTML with the nunjucks syntax for corresponding components. However, there are a few issues with this potential approach:

1. I don't know if it's possible to apply 2 template engines to a page being rendered in Spring - if it is, it's not widely used and support for it is minimal.  
2. Rendering a page twice could have a significant impact on performance (maybe it'd be acceptable, but we'd have to get it working first to see).
3. Mixing two templating engine syntaxes could lead to some messy looking pages - not great for onboarding new devs.
4. Support for rendering with nunjucks in Spring is minimal (in theory it's possible, but I haven't been able to get it working yet).


Issue 4 is of interest to me at the moment, if we can resolve it and get Spring rendering pages with nunjucks it opens up a possible alternative way of utilising the toolkit which entails replacing ThymeLeaf in Spring with nunjucks.  This would mean updating pages' syntax, but in theory nunjucks can do everything ThymeLeaf can so shouldn't be a huge problem for existing projects. 

Being able to use nunjucks in Spring requires us to use Nashorn (the JS rendering engine in JVM) to run a JS script to render the pages, so we would just need to provide the nunjucks JS script (alongside the toolkit nunjuck files) in the correct manner. There are [older guides on how to do this with Mustache](https://mkyong.com/spring-mvc/spring-mvc-mustache-js-template-example/), so in theory it should work with nunjucks. 

A complication of replacing ThymeLeaf entirely is that we can't use a spring-boot library to set up the template engine (like ThymeLeaf has) which means we'll need to add a lot of additional code ourselves to configure things like the security settings relating to serving-up the resources which is normally hidden away - we can hopefully wrap all of it up in a library though (which could also provide the required toolkit and nunjucks resources)  Switching from ThymeLeaf to nunjucks could be a good idea given how ThymeLeaf's usage in UK Gov frontends has [decreased over the last few years](https://github.com/alphagov/government-frontend-survey/blob/master/results-2021.md).

Whether we replace ThymeLeaf with nunjucks or get them working alongside each other, updating the components would be as easy as pulling the latest nunjucks files from the toolkit repo. This would be far easier than trying to convert them into ThymeLeaf fragments (either through some automated approach or manually) so would be my preferred way of doing it - if it's possible.

**Comment by yantantether** [comment]
On 1-4, I'm not sure that mixing syntaxes, or double rendering is a clean approach. I think we should stick with Thymeleaf as the standard syntax for Spring Boot UI applications. Its the default, and Spring FE developers know how it works. A port to Thymeleaf would have to be just that, client app developers should use Thymeleaf syntax and nothing else.

I think that leaves two options:

## Option 1 - transpile from nunjucks to thymeleaf

I agree it may not be so simple.  But...

A quick google shows we can use the [parser](https://github.com/mozilla/nunjucks/blob/master/nunjucks/src/parser.js) from core Nunjucks. This can parse a file and provide an Abstract Syntax Tree (AST). There's a nice explorer UI here that shows what we would get to work with:

* https://ogonkov.github.io/nunjucks-ast-explorer/

A further google showed up this PoC project to transpile Gov FE toolkit into twirl:

* https://github.com/hmrc/nunjucks-twirl-transpiler/tree/master

It is complicated, and I'm not sure how we'd convert features like filters into Thymeleaf. It could be worth pursuing.

## Option 2 - maintain handcrafted Thymeleaf templates

I think this is the natural extension of having someone notify projects of changes within the HTML, described above. We will make it easier to propagate changes, as we just have to release a new version inline with the NHS frontend lib. But it'd be a manual process to track what's changed and apply it to the thymleaf codebase.

We could make this easier on ourselves, by structuring the Thymleaf code in the same way as the nunjucks. E.g. looking at the [Header](https://github.com/nhsuk/nhsuk-frontend/tree/main/packages/components/header) component, there's quite a few nested templates. If we can align to the same structure, applying changes should be simpler.

---

How about we try this in stages, and if its not working out, we fail fast:

1. Create a PoC project with basic elements. Align to the nunjucks structure as much as possible but not get too hung up about it
2. If this works, then we could see how best to structure it for ongoing releases
3. If that works, then we could look at transpiling to make it even more efficient

**Reply by nhsbsa-jamcn** [reply]
This sounds great to me. Getting a single source of the toolkit components for ThymeLeaf projects is a great starting point. Then finding ways we can reduce the overhead of updating the fragments can be an ongoing bit of work.

It's probably best if we start off with a fresh PoC project, My implementation of the fragments I've added so far are likely to be a bit too restrictive for general use.  However, the structure of the project (pom, jgitver, resources, etc) may prove to be a useful base to work from as we know it works. I'm currently looking at adding Java classes to support the component fragments, which could be useful to port over too.

**Reply by aircwo** [reply]
Is Tymeleaf actively chosen as part of the tech stack within new projects? 
Would a transpiler / template-like framework be for existing projects?

**Reply by yantantether** [reply]
@WORTI3 I think you make two very goodpoints which we should discuss:

## Java/Spring/Thymeleaf for frontends

We don't have a policy on which tech stack we use for frontends. There are clear advantages to using Node as it aligns to both Gov and NHS toolkits. We've come a long way in our maturity in Node development too, but there's also room for improvement.

I think the main reason project teams have for choosing Java is familiarity and skills. Most teams are very Java biased, so its a natural choice, and reduces immediate risk. Long term I'm not so sure.

## Adoption in existing projects

To migrate over to a new Thymeleaf toolkit would take significant effort. I don't think we should underestimate this.

## Maintenance

Another one we should consider before we go too far. If we develop a Thymeleaf port, then we need an active maintenance team to keep on top of NHS frontend releases, and pushing out updates.


**Maybe a FE community meetup?**

**Reply by aircwo** [reply]
Are either options maintainable in the long-term? 
Migrating to a new toolkit would be a significant shift and would be hard to fit into a projects tight deadlines / re-skill devs to switch and use the toolkit. Bugs and issues raised could then have a nock-on impact if the toolkit isn't backed by a dedicated team. Same goes for the porting tool & repo. 
We would also need to make sure whatever is produced is nothing like past failings.. LIS YAFF. 
Think ultimately it comes down to the individual projects and a decision like this should come from their dev leads?
